fix #6745 feat(nimbus): Use a multi-select for feature ids

brennie opened the pull request:
Because

- we are moving to a multi-feature world;
- multifeature support has not been QA'd on all platforms yet (see
  [QA-1977][1]); and

this commit

- updates the branches page to use a multi-select for features; and
- limits the selection to a single feature, which can be undone once QA
  has finished.

[1]: https://mozilla-hub.atlassian.net/browse/QA-1977

brennie commented:
Blocked on #8860

jaredlockhart reviewed (commented):
Got into a couple interesting states by triggering some errors, then adding and removing features, the list of values gets out of sync with the list of errors in memory.  

<img width="1182" alt="image" src="https://github.com/mozilla/experimenter/assets/119884/96dbd710-b608-41ad-b57b-0b8bec60db8d">

Refresh page:

<img width="1180" alt="image" src="https://github.com/mozilla/experimenter/assets/119884/22522a0a-0621-496b-806d-b00af1d6ac32">


I think we have to assert some kind of stable ordering here, like sorting by feature slug?  That way we don't get the values and errors lists out of sync?  Or maybe when the features list changes we should just throw out the existing errors?

jaredlockhart commented:
Okay!  Testing locally things are looking really goooooodddd

I noticed the summary page doesn't show the feature names:

<img width="1220" alt="image" src="https://github.com/mozilla/experimenter/assets/119884/be6148bd-2452-40aa-ac34-85a90697ccd8">

Do you wanna handle that in here or do it as a followup?

jaredlockhart commented on experimenter/experimenter/nimbus-ui/src/components/Summary/TableBranches/index.tsx:
```diff
@@ -190,12 +190,23 @@ const TableBranch = ({
+        {featureValues ? (
+          featureValues.map((featureValue, idx) => (
+            <tr key={idx}>
+              <th>Value</th>
```
Right here could just be

<feature name> Value

jaredlockhart replied:
And probably stable ordered by feature name/slug?

brennie commented:
I updated the API to always return the feature values ordered by the feature config ID. I also updated the frontend select component and input values to be ordered the same.

brennie commented:
I updated the v5 serializer to ensure that it always orders values by their feature config IDs, just like the graphQL api.

jaredlockhart commented on experimenter/experimenter/experiments/api/v5/serializers.py:
```diff
@@ -379,6 +379,21 @@ class Meta:
         )
 
 
+class NimbusBranchFeatureValueListSerializer(serializers.ListSerializer):
+    def to_representation(self, data):
+        """Return a list of primitive data types representing the objects
+
+        This enforces that the serialized data is ordered by the feature config ID.
+        """
+        super().to_representation
+        iterable = data.all() if isinstance(data, models.Manager) else data
+
+        return [
+            self.child.to_representation(item)
+            for item in iterable.order_by("feature_config__id")
+        ]
```
Yeah exactly.  Not obvious, kind of a hassle.  But yep.

jaredlockhart commented on experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_experiment_serializer.py:
```diff
@@ -327,15 +327,25 @@ def test_saves_branches_single_feature(self):
-                "feature_value": "",
+                "feature_values": [
+                    {
+                        "feature_config": feature_config.id,
+                        "value": "",
+                    },
+                ],
             },
             "treatment_branches": [
                 {
                     "name": "treatment",
                     "description": "a treatment",
                     "ratio": 1,
-                    "feature_value": "{'this': 'that'}",
-                }
+                    "feature_values": [
+                        {
+                            "feature_config": feature_config.id,
+                            "value": "{'this': 'that'}",
+                        }
+                    ],
+                },
```
Might be worth putting two features in there and asserting their presence and ordering?

jaredlockhart reviewed (approved):
Okay wow.  Can't believe it's finally here.  Tested locally, everything was A++ 11/10.  Great job.  Form works fantastically, summary page has names, recipe serializer includes the values properly.  I think it all looks right.  Only two things jumping out at me are
* Could cover the feature value ordering in a test somewhere
* Still think we might wanna nuke the warnings/errors if the feature config selector changes

But nothing blocking and things we can do as followups if we want.

Thank you @brennie this is gonna be huge 🎉 🎉 🎉 🎉 🎉

jaredlockhart reviewed (dismissed):
Oh now I'm not getting any errors?

<img width="1277" alt="image" src="https://github.com/mozilla/experimenter/assets/119884/b107fa9f-25d5-44bc-b1fd-2c4982c8876a">


<img width="1564" alt="image" src="https://github.com/mozilla/experimenter/assets/119884/acdc3383-374e-4a4f-a0ba-20ddb2ba1a56">

After clicking save.

Refreshing the page brings them back

<img width="1537" alt="image" src="https://github.com/mozilla/experimenter/assets/119884/d7524c12-bc58-427c-b273-3b6961c65c20">


Errors/warnings should be shown immediately after clicking save.

jaredlockhart reviewed (approved):
Okay.  Ran through it all again.  THIS is the best experience it's been so far.  I never got into a confused state.  Resetting the errors made a big difference.  Now the UI state is always consistent.  All the validation is all working in all directions, every combination I could think of.  Thank you @brennie for getting this over the finish line.  Now let's land this thing!! 🎉 🎉 🎉 🎉

yashikakhurana commented:
wooo hoooo can't wait to see this landing 🎉 , great job  @brennie 👏

brennie commented:
Blocked on QA

jaredlockhart commented:
Let's block on this https://mozilla-hub.atlassian.net/browse/EXP-3608

brennie commented:
Update: I've removed the parts of this that change the API responses and have limited the frontend to only allow selecting one feature at a time.

brennie commented:
https://github.com/mozilla/experimenter/pull/8907/files#diff-99544655f6d0c765eebe36ad77f4df48ff10ab000d88ac14a84e25645aef56f6R208-R209 contains the logic for limiting the frontend to a single value. It is accompanied by a test. Removal of that line will allow for multi-feature frontend.

jaredlockhart commented on experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/index.tsx:
```diff
@@ -176,11 +201,15 @@ export const FormBranches = ({
+    setIsSelectValid(newValue.length > 0);
+    setSelectDirty(true);
+
+    // Limit to a single value until multi-feature support has been QA'd.
+    newValue = newValue.length > 0 ? [newValue[newValue.length - 1]] : newValue;
```
👍

jaredlockhart reviewed (approved):
Okay yep, tested locally, everything looks right.  Funny, always chop off the selected list to the last new entry.

I was thinking of this like on the metrics page:

https://github.com/mozilla/experimenter/blob/main/experimenter/experimenter/nimbus-ui/src/components/PageEditMetrics/FormMetrics/index.tsx#L163

But this works too.

Awesome let's get this landed 🎉 🎉 🎉
